components/buildList/js/buildList.controller.js

```diff
@@ -6,6 +6,464 @@ buildApp.controller('buildListController', ['$scope', '$q', '$interval', 'buildL
+			var jDefered = $q.defer();
+			jobsDeferals.push(jDefered.promise);
+			updateJob(job, jDefered);
+		});
+
+		//update firewall information after all its jobs have been updated
+		$q.all(jobsDeferals).then(function() {
+
+			angular.forEach(firewall.jobs, function(job) {
+				jobsStates.push(job.status);
+			});
+
+			firewall.status = setStatusOfProcess(jobsStates);
+			firewallDefered.resolve(); 
+		});
+	}
+
+	/**
+	 * Updates the build status
+	 * @method updateBuild
+	 * @param  {Object} build Build that is analysed, for each build all its firewalls are requested to be updated
+	 * @param  {Object} firewallDefered
+	 */
+	function updateBuild(build, buildDefered) {
+
+		if (build.status === 'complete' || build.status === 'rejected') {
+			buildDefered.resolve();
+			return;
+		}
+
+		var firewallsPromises = [];
+
+		angular.forEach(build.firewalls, function(firewall) {
+			var firewallDefered = $q.defer();
+			firewallsPromises.push(firewallDefered.promise);
+			updateFirewall(firewall, firewallDefered); 
+		});
+
+		//update build information after all its firewalls have been updated
+		$q.all(firewallsPromises).then(function() {
+			var fStates = [];
+
+			var buildJobs = {
+				metrics: {
+					percentage: 0,
+					status: []
+				},
+				build: {
+					percentage: 0,
+					status: []
+				},
+				unitTest: {
+					percentage: 0,
+					status: []
+				},
+				functionalTest: {
+					percentage: 0,
+					status: []
+				}
+			};
+
+			angular.forEach(build.firewalls, function(firewall) {
+				fStates.push (firewall.status);
+				angular.forEach(firewall.jobs, function(job, jobName) {
+					buildJobs[jobName].percentage += job.percentage;
+					buildJobs[jobName].status.push(job.status);
+				});
+			});
+
+			angular.forEach(buildJobs, function(value) {
+				value.percentage = parseInt(value.percentage / build.firewalls.length, 10);
+				value.status = setStatusOfProcess(value.status);
+			});
+
+			build.jobs = buildJobs;
+			build.status = setStatusOfProcess(fStates);
+			buildDefered.resolve();
+		});
+	}
+
+
+	/**
+	 * Starts to calculate the states of the builds.
+	 * @method updateBuildsList
+	 */
+	function updateBuildsList() {
+
+		timeElapsed += intervalTime;
+		var overallStatus = 'running';
+		var buildPromises = [];
+
+		angular.forEach($scope.buildList, function(build) {
+			var buildDefered = $q.defer();
+			buildPromises.push(buildDefered.promise);
+			updateBuild(build, buildDefered);
+		});
+
+		$q.all(buildPromises).then(function() {
+			var buildStates = [];
+
+			angular.forEach($scope.buildList, function(build) {
+				buildStates.push (build.status);
+			});
+
+			overallStatus = setStatusOfProcess(buildStates);
+			if (overallStatus === 'complete' || overallStatus === 'rejected') {
+				$interval.cancel(updateBuildsInterval);
+			}
+		});
+	}
+
+	init();
+}]).controller('buildListController', ['$scope', '$q', '$interval', 'buildListService', function ($scope, $q, $interval, buildListService) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This function has 239 lines, which is greater than the 100 lines authorized. Split it into smaller functions. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AS138)